Add lookup name to iterable lookup

camuthig opened the pull request:
This name allows for the in lookup to behave normally in things like django-filter FilterSet logic. Without this value, the in lookup will appear as if it were an exact lookup to other code.

camuthig committed:
Add lookup name to iterable lookup

This name allows for the in lookup to behave normally in things like
django-filter FilterSet logic. Without this value, the in lookup will
appear as if it were an exact lookup to other code.

carltongibson commented:
This probably needs a test case to show the issue fixed, and prevent regressions.

nshafer commented:
Can you give me an example of exactly what problem you're trying to solve? I've done some tests with django-filter, and it seems to work fine for me. E.g.

```python
class ReferenceInFilter(BaseInFilter, CharFilter):
    pass


class BookFilter(FilterSet):
    reference_id = django_filters.CharFilter(field_name='reference_id', lookup_expr='iexact')
    reference_id_in = ReferenceInFilter(field_name='reference_id', lookup_expr='in')

    class Meta:
        model = Book
        fields = {
            'id': ['exact'],
            'name': ['exact', 'icontains'],
            'author': ['exact'],
        }
```

all works as expected.

camuthig commented:
I will work on adding some tests soon.

As for use case, I am working with DRF and DRF Spectacular with `HashidAutoField`, without integer lookup support. 

```python
class Blog(models.Model):
    id = HashidAutoField(primary_key=True, prefix='blog_', min_length=20, allow_int_lookup=False)
    title = models.CharField(max_length=255)
```

I noted that if I created a filterset like

```python
class BlogFilterSet(FilterSet):
    class Meta:
        model = Blog
        fields = {
            'id': ['in'],
        }
```

then my generated documentation via DRF Spectacular came out incorrect

<img width="651" alt="Screen Shot 2022-12-22 at 6 00 31 PM" src="https://user-images.githubusercontent.com/5178217/209256017-e6a8734b-8c8e-4bd7-a05d-946091884a75.png">

This is because DRF Spectacular and django-filter are using this lookup name to figure out the correct type, and under the hood create a `ConcreteInFilter` class. Once the lookup name is corrected, the documentation generation was resolved.

<img width="618" alt="Screen Shot 2022-12-22 at 6 03 34 PM" src="https://user-images.githubusercontent.com/5178217/209256316-6b8d48e8-2de4-4040-9d89-6e31fbba460d.png">

camuthig committed:
Add explicit exact lookup name

During testing of the in lookup name for the hashid lookup type, it was
determined the explicit behaviors around the exact lookup type were also
not defined. In something like django-filters, this could make for odd
behaviors depending on how the project defines the `DEFAULT_LOOKUP_EXPR`
setting. It defaults to exact, but if set to something else then the
exact lookup may not behave as expected.

camuthig committed:
Add tests for lookup name behaviors

camuthig commented:
I have added some tests based on how django-filter expects lookup names to behave. I added the name value to the `Exact` lookup as well based on these tests.
